sympy/calculus/finite_diff.py

```diff
@@ -452,32 +450,54 @@ def differentiate_finite(expr, *symbols,
     >>> differentiate_finite(f(x)*g(x), x, points=[x-h, x+h])
     -f(-h + x)*g(-h + x)/(2*h) + f(h + x)*g(h + x)/(2*h)
 
-    Note that the above form preserves the product rule in discrete form.
-    If we want we can pass ``evaluate=True`` to get another form (which is
-    usually not what we want):
+    Note that the above form preserves the product rule in discrete form. If we
+    want we can pass ``evaluate=True`` to get another form (which is usually
+    not what we want):
 
     >>> differentiate_finite(f(x)*g(x), x, points=[x-h, x+h], evaluate=True).simplify()
     -((f(-h + x) - f(h + x))*g(x) + (g(-h + x) - g(h + x))*f(x))/(2*h)
 
-    ``differentiate_finite`` works on any expression:
+    ``differentiate_finite`` works on any expression, including expressions
+    with embedded derivatives that don't support ``evaluate=True``:
 
     >>> differentiate_finite(f(x) + sin(x), x, 2)
     -2*f(x) + f(x - 1) + f(x + 1) - 2*sin(x) + sin(x - 1) + sin(x + 1)
     >>> differentiate_finite(f(x) + sin(x), x, 2, evaluate=True)
     -2*f(x) + f(x - 1) + f(x + 1) - sin(x)
     >>> differentiate_finite(f(x, y), x, y)
     f(x - 1/2, y - 1/2) - f(x - 1/2, y + 1/2) - f(x + 1/2, y - 1/2) + f(x + 1/2, y + 1/2)
+    >>> differentiate_finite(f(x)*g(x).diff(x), x)
+    (-g(x) + g(x + 1))*f(x + 1/2) - (g(x) - g(x - 1))*f(x - 1/2)
+
+    To make finite difference with non-constant discretization step use
+    undefined functions:
+
+    >>> dx = Function('dx')
+    >>> differentiate_finite(f(x)*g(x).diff(x), points=dx(x))
+    -(-g(x - dx(x)/2 - dx(x - dx(x)/2)/2)/dx(x - dx(x)/2) +
+    g(x - dx(x)/2 + dx(x - dx(x)/2)/2)/dx(x - dx(x)/2))*f(x - dx(x)/2)/dx(x) +
+    (-g(x + dx(x)/2 - dx(x + dx(x)/2)/2)/dx(x + dx(x)/2) +
+    g(x + dx(x)/2 + dx(x + dx(x)/2)/2)/dx(x + dx(x)/2))*f(x + dx(x)/2)/dx(x)
 
     """
     # Key-word only arguments only available in Python 3
     points = kwargs.pop('points', 1)
     x0 = kwargs.pop('x0', None)
     wrt = kwargs.pop('wrt', None)
     evaluate = kwargs.pop('evaluate', False)
+    if any(term.is_Derivative for term in list(preorder_traversal(expr))):
+        evaluate = False
```

Comment:
When is evaluate ever set to True. It seems you set to it to False if you have any term that does not have a derivative.

Reply:
I don't really see the use case when one wants to set it True and that is stated in the documentation as well  

>    Note that the above form preserves the product rule in discrete form. If we
>    want we can pass ``evaluate=True`` to get another form (which is usually
>    not what we want):

However, I decided to leave because maybe someone is relying on it and I didin't want to beak backwards compatibility. This means, if needed, one can still use `evaluate=True` for expressions without derivatives inside them

```
In[3]: differentiate_finite(f(x)*g(x), x, evaluate=True)
Out[3]: (-f(x - 1/2) + f(x + 1/2))*g(x) + (-g(x - 1/2) + g(x + 1/2))*f(x)

In [4]: differentiate_finite(f(x)*g(x), x, evaluate=False)                                                                                                           
Out[4]: -f(x - 1/2)*g(x - 1/2) + f(x + 1/2)*g(x + 1/2)
```

Reply:
Is the `evaluate=True` behaviour actually useful or should it be deprecated?

Reply:
I didn't want to open this discussion or make that call but since you mentioned it... I don't see `evaluate=True` so useful... Expressions passed to `differentiate_finite` are usually in the form in which they should be discretized, I don't think this function should manipulate expressions before discretization in any way.

Reply:
@bjodah What is the purpose of `evaluate=True` in `differentiate_finite`. It seems to be discouraged in the docstring. Should it be deprecated and removed?

Reply:
I can't remember I'm afraid. I think we introduced it for backwards compatibility reasons. I'm fine with deprecating it.

Reply:
@vezeli can you add a deprecation warning for the `evaluate` flag and update the docstring?

Reply:
@oscarbenjamin It seems like the PR is in line with the wisdom of [Zen of SymPy](https://github.com/sympy/sympy/blob/master/sympy/this.py):

> Unevaluated is better than evaluated.

Additionally, I changed the docstrings in other functions that were not correctly leveled.